### DISCUSSION TITLE
Track protected via HTTP Basic Authentication

### DISCUSSION THREAD

**Discussion by laceysanderson** [GMOD/jbrowse-components#4299] [Q&A] [answered]
I am seeing a lot of mention in the issue queue that implies that support for this already exists but I cannot find any information on how to accomplish it 🙈 

## What we want to accomplish

We want to update our current JBrowse2 config to only include URLs (every track is behind HTTP Basic auth via htaccess) for two reasons:
1. to separate our data files from our JBrowse for less duplication.
2. to allow us to clone the JBrowse locally (all files and config are in a private git repo) and test it there.

## How I think we want to implement it

I've seen mention of JBrowse2 having functionality called "internet accounts" which can describe how to access third party resources. I've also seen mention of an "HTTP Basic" InternetAccountType but I can't seem to find documentation that ties it all together and tells me how to do this 😅 

## Resources

This is some of what I've been coming across but none specifically tell me how to do this 🙈 
- Issue/Task stream of thought on implementing access control in JBrowse: Issue #604 
- Possible limitation of HTTP Basic #3748
- [Sourceforge discussion that mentions track authentication](https://sourceforge.net/p/gmod/mailman/message/37855146/) but it ended up not being what the OP needed
- Issue #3105 mentioning it would be good to have documentation showing how to do this 😉

**Comment by cmdcolin** [comment, marked as answer]
here is some possible help

1) you can configure an HTTPBasicInternetAccount in your config.json https://github.com/GMOD/jbrowse-components/blob/42442fedf2c1e25f70df3cf8ec1e2d8ba5106cf2/test_data/volvox/config_auth.json#L237-L242

2) you can make every "location" object have that internetAccountId added to it example https://github.com/GMOD/jbrowse-components/blob/main/test_data/volvox/config_auth.json#L188-L209

if you want to run the config_auth demo to try it out yourself run these steps


```
git clone https://github.com/GMOD/jbrowse-component
cd jbrowse-components
yarn
cd products/jbrowse-web
yarn start-basicauth-server ## this is a simple  "express.js" dev server that has http basic auth on it

## then in another tab
cd products/jbrowse-web
yarn start

## then in browser
visit http://localhost:3000/?config=test_data%252Fvolvox%252Fconfig_auth.json
```

then when you select a basic auth track from the tracklist, it will prompt for login which is admin:password

![image](https://github.com/GMOD/jbrowse-components/assets/6511937/824e79b7-681d-4d69-92ad-e08f4e4bc8d9)

**Reply by laceysanderson** [reply]
Awesome! Thank you for the quick and detailed answer!!! This looks really straight forward and intuitive 😃 @carolyncaron can you give this a try on our instance?

**Reply by carolyncaron** [reply]
So I've confirmed that this has worked wonderfully for our instance, including when I add the internetAccountId to the reference sequence track. :tada:

I've run into a problem with Multi Quantitative tracks however. Since there's no location object, I don't know where to add the internetAccountId for it. Is it possible to add authorization to these tracks?

**Reply by cmdcolin** [reply]
it is more verbose, but I believe if you use the 'subadapters' array rather than the 'bigWig' array of urls, then you can add the internetAccountId to each individual subadapter

https://jbrowse.org/jb2/docs/config_guides/multiquantitative_track/#general-multiquantitativetrack-options

example test I tried and should work

```
{
      "type": "MultiQuantitativeTrack",
      "trackId": "volvox_microarray_multi",
      "name": "MultiWig",
      "assemblyNames": ["volvox"],
      "adapter": {
        "type": "MultiWiggleAdapter",
        "subadapters": [
          {
            "name": "k1",
            "source": "j1",
            "type": "BigWigAdapter",
            "bigWigLocation": {
              "uri": "http://localhost:3040/data/v1.cram.bw",
              "internetAccountId": "customBasicAuth"
            }
          },
          {
            "name": "k2",
            "source": "j2",
            "type": "BigWigAdapter",
            "bigWigLocation": {
              "uri": "http://localhost:3040/data/v2.cram.bw",
              "internetAccountId": "customBasicAuth"
            }
          },
          {
            "name": "k3",
            "type": "BigWigAdapter",
            "bigWigLocation": {
              "uri": "http://localhost:3040/data/v3.cram.bw",
              "internetAccountId": "customBasicAuth"
            }
          },
          {
            "name": "k4",
            "type": "BigWigAdapter",
            "bigWigLocation": {
              "uri": "http://localhost:3040/data/v4.cram.bw",
              "internetAccountId": "customBasicAuth"
            }
          }
        ]
      }
    },

```

**Reply by carolyncaron** [reply]
Thanks a lot Colin! I confirmed this works for us :+1:
